File: workshop/webGL/3/main/1 - core/shapes/character.js
this.character = function(){
    var self = this;

    //attributes 
        //protected attributes
            const type = 'character'; this.getType = function(){return type;}
            const vectorLibrary = library.character.vectorLibrary;
            const defaultFontName = 'defaultThin';

        //simple attributes
            this.name = '';
            this.parent = undefined;
            this.dotFrame = false;
            this.extremities = { points:[], boundingBox:{} };
            this.ignored = false;
            this.colour = {r:1,g:0,b:0,a:1};
        //advanced use attributes
            this.devMode = false;
            this.stopAttributeStartedExtremityUpdate = false;

        //attributes pertinent to extremity calculation
            var x = 0;                  this.x =      function(a){ if(a==undefined){return x;}      x = a;      if(this.devMode){console.log(this.getAddress()+'::x');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var y = 0;                  this.y =      function(a){ if(a==undefined){return y;}      y = a;      if(this.devMode){console.log(this.getAddress()+'::y');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var angle = 0;              this.angle =  function(a){ if(a==undefined){return angle;}  angle = a;  if(this.devMode){console.log(this.getAddress()+'::angle');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var anchor = {x:0,y:0};     this.anchor = function(a){ if(a==undefined){return anchor;} anchor = a; if(this.devMode){console.log(this.getAddress()+'::anchor');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var width = 10;             this.width =  function(a){ if(a==undefined){return width;}  width = a;  if(this.devMode){console.log(this.getAddress()+'::width');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var height = 10;            this.height = function(a){ if(a==undefined){return height;} height = a; if(this.devMode){console.log(this.getAddress()+'::height');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var scale = 1;              this.scale =  function(a){ if(a==undefined){return scale;}  scale = a;  if(this.devMode){console.log(this.getAddress()+'::scale');} if(this.stopAttributeStartedExtremityUpdate){return;} computeExtremities(); };
            var font = defaultFontName;
                this.font = function(a){
                    if(a==undefined){return font;}
                    if(this.devMode){console.log(this.getAddress()+'::font');} 

                    font = vectorLibrary[a] == undefined ? defaultFontName : a;

                    if(this.devMode){console.log(this.getAddress()+'::font - isLoaded:',vectorLibrary[font].isLoaded);} 
                    if(!vectorLibrary[font].isLoaded){ setTimeout(function(){ self.font(font); },100,font); }

                    producePoints();

                    if(this.stopAttributeStartedExtremityUpdate){return;} 
                    computeExtremities(); 
                };
            var character = '';
                this.character = function(a){
                    if(a==undefined){return character;} 
                    if(this.devMode){console.log(this.getAddress()+'::character - '+a);}

                    character = a; 
                    producePoints();

                    if(this.stopAttributeStartedExtremityUpdate){return;} 
                    computeExtremities(); 
                };
            var printingMode = {
                horizontal:'left', //left / middle / right
                vertical:'bottom', //top  / middle / bottom
            };
                this.printingMode = function(a){
                    if(a==undefined){return printingMode;} 
                    printingMode = {
                        horizontal: a.horizontal != undefined || a.horizontal != '' ? a.horizontal : printingMode.horizontal,
                        vertical: a.vertical != undefined || a.vertical != '' ? a.vertical : printingMode.vertical,
                    };

                    producePoints();

                    if(this.stopAttributeStartedExtremityUpdate){return;} 
                    computeExtremities(); 
                };

    //addressing
        this.getAddress = function(){ return (this.parent != undefined ? this.parent.getAddress() : '') + '/' + this.name; };

    //character data
        this.top = function(){ return vectorLibrary[font][character] == undefined ? 0 : vectorLibrary[font][character].top; };
        this.bottom = function(){ return vectorLibrary[font][character] == undefined ? 1 : vectorLibrary[font][character].bottom; };
        this.left = function(){ return vectorLibrary[font][character] == undefined ? 0 : vectorLibrary[font][character].left; };
        this.right = function(){ return vectorLibrary[font][character] == undefined ? 1 : vectorLibrary[font][character].right; };
        function producePoints(){            
            points = (vectorLibrary[font][character] == undefined ? vectorLibrary[font]['default'].vector : vectorLibrary[font][character].vector).concat([]); //the concat, differentiates the point data

            //adjust for vertical printingMode
                var horizontalAdjust = vectorLibrary[font][character] == undefined ? 0 : vectorLibrary[font][character].right;
                if(printingMode.horizontal == 'middle'){ horizontalAdjust = horizontalAdjust/2; }
                if(printingMode.horizontal != 'left'){
                    for(var a = 0; a < points.length; a+=2){
                        points[a] = points[a] - horizontalAdjust;
                    }
                }

            //adjust for horizontal printingMode
                var verticalAdjust = vectorLibrary[font][character] == undefined ? 0 : vectorLibrary[font][character].top;
                if(printingMode.vertical == 'middle'){ verticalAdjust = verticalAdjust/2; }
                if(printingMode.vertical != 'bottom'){
                    for(var a = 0; a < points.length; a+=2){
                        points[a+1] = points[a+1] - verticalAdjust;
                    }
                }

            pointsChanged = true;
        }

    //webGL rendering functions
        var pointsChanged = true;
        var points = [ 0,0, 1,0, 1,1,  0,0, 1,1, 0,1 ];
        var vertexShaderSource = 
            _canvas_.library.gsls.geometry + `
            //constants
                attribute vec2 point;

            //variables
                struct location{
                    vec2 xy;
                    float scale;
                    float angle;
                };
                uniform location adjust;

                uniform vec2 resolution;
                uniform vec2 dimensions;
                uniform vec2 anchor;

            void main(){
                //using the 'adjust' values; perform anchored rotation, and leave shape with it's anchor over the chosen point
                    vec2 P = dimensions * adjust.scale * (point - anchor);
                    P = vec2( P.x*cos(adjust.angle) + P.y*sin(adjust.angle), P.y*cos(adjust.angle) - P.x*sin(adjust.angle) ) + adjust.xy;

                //convert from unit space to clipspace
                    gl_Position = vec4( (((P / resolution) * 2.0) - 1.0) * vec2(1, -1), 0, 1 );
            }
        `;
        var fragmentShaderSource = `  
            precision mediump float;
            uniform vec4 colour;
                                                                        
            void main(){
                gl_FragColor = colour;
            }
        `;
        var point = { buffer:undefined, attributeLocation:undefined };
        var uniformLocations;
        function updateGLAttributes(context,adjust){
            //buffers
                //points
                    if(point.buffer == undefined || pointsChanged){
                        point.attributeLocation = context.getAttribLocation(program, "point");
                        point.buffer = context.createBuffer();
                        context.enableVertexAttribArray(point.attributeLocation);
                        context.bindBuffer(context.ARRAY_BUFFER, point.buffer); 
                        context.vertexAttribPointer( point.attributeLocation, 2, context.FLOAT,false, 0, 0 );
                        context.bufferData(context.ARRAY_BUFFER, new Float32Array(points), context.STATIC_DRAW);
                        pointsChanged = false;
                    }else{
                        context.bindBuffer(context.ARRAY_BUFFER, point.buffer); 
                        context.vertexAttribPointer( point.attributeLocation, 2, context.FLOAT,false, 0, 0 );
                    }

            //uniforms
                if( uniformLocations == undefined ){
                    uniformLocations = {
                        "adjust.xy": context.getUniformLocation(program, "adjust.xy"),
                        "adjust.scale": context.getUniformLocation(program, "adjust.scale"),
                        "adjust.angle": context.getUniformLocation(program, "adjust.angle"),
                        "resolution": context.getUniformLocation(program, "resolution"),
                        "dimensions": context.getUniformLocation(program, "dimensions"),
                        "anchor": context.getUniformLocation(program, "anchor"),
                        "colour": context.getUniformLocation(program, "colour"),
                    };
                }

                context.uniform2f(uniformLocations["adjust.xy"], adjust.x, adjust.y);
                context.uniform1f(uniformLocations["adjust.scale"], adjust.scale);
                context.uniform1f(uniformLocations["adjust.angle"], adjust.angle);
                context.uniform2f(uniformLocations["resolution"], context.canvas.width, context.canvas.height);
                context.uniform2f(uniformLocations["dimensions"], width, height);
                context.uniform2f(uniformLocations["anchor"], anchor.x, anchor.y);
                context.uniform4f(uniformLocations["colour"], self.colour.r, self.colour.g, self.colour.b, self.colour.a);
        }
        var program;
        function activateGLRender(context,adjust){
            if(program == undefined){ program = core.render.produceProgram(self.getType(), vertexShaderSource, fragmentShaderSource); }

            context.useProgram(program);
            updateGLAttributes(context,adjust);

            context.drawArrays(context.TRIANGLES, 0, points.length/2);
        }

    //extremities
        function computeExtremities(informParent=true,offset){
            if(self.devMode){console.log(self.getAddress()+'::computeExtremities');}

            //get offset from parent, if one isn't provided
                if(offset == undefined){ offset = self.parent && !self.static ? self.parent.getOffset() : {x:0,y:0,scale:1,angle:0}; }    
            //calculate adjusted offset based on the offset
                var point = _canvas_.library.math.cartesianAngleAdjust(x,y,offset.angle);
                var adjusted = { 
                    x: point.x*offset.scale + offset.x,
                    y: point.y*offset.scale + offset.y,
                    scale: offset.scale*scale,
                    angle: -(offset.angle + angle),
                };
            //calculate points based on the adjusted offset
                self.extremities.points = [];
                for(var a = 0; a < points.length; a+=2){
                    var P = {
                        x: adjusted.scale * width * (points[a] - anchor.x), 
                        y: adjusted.scale * height * (points[a+1] - anchor.y), 
                    };

                    self.extremities.points.push({ 
                        x: P.x*Math.cos(adjusted.angle) + P.y*Math.sin(adjusted.angle) + adjusted.x,
                        y: P.y*Math.cos(adjusted.angle) - P.x*Math.sin(adjusted.angle) + adjusted.y,
                    });
                }
                self.extremities.boundingBox = _canvas_.library.math.boundingBoxFromPoints(self.extremities.points);

            //// "point in poly" detection currently doesn't understand polys with holes in them, so these complex
            //// shapes are being simplified to their boudning boxes
                self.extremities.points = [
                    {x:self.extremities.boundingBox.topLeft.x,y:self.extremities.boundingBox.topLeft.y},
                    {x:self.extremities.boundingBox.bottomRight.x,y:self.extremities.boundingBox.bottomRight.y},
                ];

            //if told to do so, inform parent (if there is one) that extremities have changed
                if(informParent){ if(self.parent){self.parent.updateExtremities();} }
        }
        this.computeExtremities = computeExtremities;

    //lead render
        function drawDotFrame(){
            //draw shape extremity points
                self.extremities.points.forEach(a => core.render.drawDot(a.x,a.y));
            //draw bounding box top left and bottom right points
                core.render.drawDot(self.extremities.boundingBox.topLeft.x,self.extremities.boundingBox.topLeft.y,3,{r:0,g:1,b:1,a:0.5});
                core.render.drawDot(self.extremities.boundingBox.bottomRight.x,self.extremities.boundingBox.bottomRight.y,3,{r:0,g:1,b:1,a:0.5});
        };
        this.render = function(context,offset={x:0,y:0,scale:1,angle:0}){
            //combine offset with shape's position, angle and scale to produce adjust value for render
                var point = _canvas_.library.math.cartesianAngleAdjust(x,y,offset.angle);
                var adjust = { 
                    x: point.x*offset.scale + offset.x,
                    y: point.y*offset.scale + offset.y,
                    scale: offset.scale*scale,
                    angle: -(offset.angle + angle),
                };

            //activate shape render code
                activateGLRender(context,adjust);

            //if requested; draw dot frame
                if(self.dotFrame){drawDotFrame();}
        };
};

File: main/1 - core/engine/rust/master/src/b_element/element/character.rs
//rust
    use std::rc::Weak;
    use std::cell::RefCell;

//wasm
    use web_sys::WebGl2RenderingContext;

//core
    use crate::a_library::{
        data_type::{
            Colour,
            Offset,
            Viewbox,
            Point,
            SimplePolygon,
            ElementType,
            PrintingModePositionHorizontal,
            PrintingModePositionVertical,
            RenderDecision,
        },
        structure::{
            WebGl2programConglomerateManager,
            ImageRequester,
            FontRequester,
        },
        wasm::from_js_sys::{
            get_js_sys_object_from_js_sys_object,
            get_value_from_object__f32,
            get_value_from_object__char,
            get_value_from_object__string,
            get_value_from_object__point,
            get_value_from_object__colour,
        },
    };
    use super::super::element::ElementTrait;
    use crate::f_stats::Stats;








static VAO_NAMES: [&'static str; 1] = [
    "point"
];
// static VAO_POINTS: [f32; ?] = [
//     - generated -
// ];
static VAO_GROUPING_SIZE: [i32; 1] = [
    2,
];
static VERTEX_SHADER_SOURCE: &'static str = "#version 300 es
    #define PI 3.141592653589793

    vec2 cartesianAngleAdjust(vec2 xy, float angle){
        if(angle == 0.0 || mod(angle,PI*2.0) == 0.0){ return xy; }
        return vec2( xy.x*cos(angle) - xy.y*sin(angle), xy.y*cos(angle) + xy.x*sin(angle) ); 
    }

    //constants
        in vec2 point;

    //variables
        struct location{
            vec2 xy;
            float scale;
            float angle;
        };
        uniform location adjust;

        uniform vec2 resolution;
        uniform vec2 dimensions;
        uniform vec2 anchor;

    void main(){
        //using the 'adjust' values; perform anchored rotation, and leave shape with it's anchor over the chosen point
            vec2 P = dimensions * adjust.scale * (point - anchor);
            P = vec2( P.x*cos(adjust.angle) + P.y*sin(adjust.angle), P.y*cos(adjust.angle) - P.x*sin(adjust.angle) ) + adjust.xy;

        //convert from unit space to clipspace
            gl_Position = vec4( (((P / resolution) * 2.0) - 1.0) * vec2(1, -1), 0, 1 );
    }
";
static FRAGMENT_SHADER_SOURCE: &'static str = "#version 300 es
    precision mediump float;
    out vec4 outputColour;
    uniform vec4 colour;
                                                                
    void main(){
        outputColour = colour;
    }
";
static UNIFORMS_TO_FIND: [&'static str; 7] = [
    "adjust.xy",
    "adjust.scale",
    "adjust.angle",
    "resolution",
    "dimensions",
    "anchor",
    "colour",
];

pub struct Character {
    //hierarchy and identity
        element_type: ElementType, 
        id: usize,
        name: String, 
        self_reference: Option<Weak<RefCell<dyn ElementTrait>>>,
        parent: Option<Weak<RefCell<dyn ElementTrait>>>,

    //attributes
        //pertinent to extremity calculation
            x: f32,
            y: f32,
            angle: f32,
            scale: f32,
            width: f32,
            height: f32,
            anchor: Point,
            font_name: String,
            character: char,
            printing_mode_horizontal: PrintingModePositionHorizontal,
            printing_mode_vertical: PrintingModePositionVertical,
        //other
            ignored: bool,
            colour: Colour,

    //switches
        dot_frame: bool,

    //computed values
        extremities: SimplePolygon,
        cached_offset: Offset,
        cached_heed_camera: bool,
        vao_id: Option<usize>,
        vao_points: Vec<f32>,
        points_changed: bool,

    //render
        is_visible: bool,
        previous_is_visible: bool,
}
impl Character {
    pub fn new(id:usize, name:String) -> Character {
        Character {
            element_type: ElementType::Character,
            id: id,
            name: name,
            self_reference: None,
            parent: None,

            x: 0.0,
            y: 0.0,
            angle: 0.0,
            scale: 1.0,
            width: 10.0,
            height: 10.0,
            anchor: Point::new(0.0,0.0),
            font_name: String::from("defaultThin"),
            character: ' ',
            printing_mode_horizontal: PrintingModePositionHorizontal::Left,
            printing_mode_vertical: PrintingModePositionVertical::Bottom,

            ignored: false,
            colour: Colour::new(1.0,0.0,0.0,1.0),

            dot_frame: false,
            
            extremities: SimplePolygon::new_default(),
            cached_offset: Offset::new_default(),
            cached_heed_camera: false,
            vao_id: None,
            vao_points: vec![],
            points_changed: true,

            is_visible: false,
            previous_is_visible: false,
        }
    }

    //attributes
        //pertinent to extremity calculation
            //width
                pub fn get_width(&self) -> &f32 { &self.width }
                pub fn set_width(&mut self, new:f32, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.width = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //height
                pub fn get_height(&self) -> &f32 { &self.height }
                pub fn set_height(&mut self, new:f32, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.height = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //anchor
                pub fn get_anchor(&self) -> &Point { &self.anchor }
                pub fn set_anchor(&mut self, new:Point, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.anchor = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //font
                pub fn get_font_name(&self) -> &str { &self.font_name }
                pub fn set_font_name(&mut self, new:String, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.font_name = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //character
                pub fn get_character(&self) -> &char { &self.character }
                pub fn set_character(&mut self, new:char, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.character = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //printing_mode_horizontal
                pub fn get_printing_mode_horizontal(&self) -> &PrintingModePositionHorizontal { &self.printing_mode_horizontal }
                pub fn set_printing_mode_horizontal(&mut self, new:PrintingModePositionHorizontal, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.printing_mode_horizontal = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //printing_mode_vertical
                pub fn get_printing_mode_vertical(&self) -> &PrintingModePositionVertical { &self.printing_mode_vertical }
                pub fn set_printing_mode_vertical(&mut self, new:PrintingModePositionVertical, font_requester:&RefCell<FontRequester>, worker:&web_sys::Worker, viewbox:&Viewbox) { 
                    self.printing_mode_vertical = new;
                    self.calculate_points(font_requester, worker, viewbox);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
        //other
            pub fn get_colour(&self) -> &Colour { &self.colour }
            pub fn set_colour(&mut self, new:Colour, _viewbox:&Viewbox) {
                self.colour = new;
                self.request_render();
            }
        //unified attribute
            pub fn set_unified_attribute(
                &mut self,
                worker: &web_sys::Worker,
                x: Option<f32>,
                y: Option<f32>,
                angle: Option<f32>,
                scale: Option<f32>,
                width: Option<f32>,
                height: Option<f32>,
                anchor: Option<Point>,
                font_name: Option<String>,
                character: Option<char>,
                printing_mode_horizontal: Option<PrintingModePositionHorizontal>,
                printing_mode_vertical: Option<PrintingModePositionVertical>,
                colour: Option<Colour>,
                font_requester: &RefCell<FontRequester>,
                viewbox: &Viewbox,
            ) {
                let mut calculate_points = false;

                if let Some(x) = x { self.x = x; }
                if let Some(y) = y { self.y = y; }
                if let Some(angle) = angle { self.angle = angle; }
                if let Some(scale) = scale { self.scale = scale; }
                if let Some(width) = width { self.width = width; calculate_points = true; }
                if let Some(height) = height { self.height = height; calculate_points = true; }
                if let Some(anchor) = anchor { self.anchor = anchor; calculate_points = true; }
                if let Some(font_name) = font_name { self.font_name = font_name; calculate_points = true; }
                if let Some(character) = character { self.character = character; calculate_points = true; }
                if let Some(printing_mode_horizontal) = printing_mode_horizontal { self.printing_mode_horizontal = printing_mode_horizontal; calculate_points = true; }
                if let Some(printing_mode_vertical) = printing_mode_vertical { self.printing_mode_vertical = printing_mode_vertical; calculate_points = true; }
                if let Some(colour) = colour { self.colour = colour; }

                if calculate_points {
                    self.calculate_points(font_requester, worker, viewbox);
                } else {
                    self.compute_extremities(true, None, None);
                }
                self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                self.request_render();
            }

    //character data
        pub fn top(&self, font_requester:&RefCell<FontRequester>) -> f32 { font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_top() }
        pub fn bottom(&self, font_requester:&RefCell<FontRequester>) -> f32 { font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_bottom() }
        pub fn left(&self, font_requester:&RefCell<FontRequester>) -> f32 { font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_left() }
        pub fn right(&self, font_requester:&RefCell<FontRequester>) -> f32 { font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_right() }
}
impl ElementTrait for Character {
    //trait requirements
        //hierarchy and identity
            fn get_element_type(&self) -> &ElementType { &self.element_type }
            fn get_id(&self) -> usize { self.id }
            fn get_name(&self) -> &String{ &self.name }
            fn set_name(&mut self, new:String) { self.name = new; }
            fn set_self_reference(&mut self, new:Weak<RefCell<dyn ElementTrait>>) { self.self_reference = Some(new); }
            fn get_parent(&self) -> &Option<Weak<RefCell<dyn ElementTrait>>> { &self.parent }
            fn set_parent(&mut self, new:Weak<RefCell<dyn ElementTrait>>) { self.parent = Some(new); }
            fn clear_parent(&mut self) { self.parent = None; }

        //position
            //x
                fn get_x(&self) -> f32 { self.x }
                fn set_x(&mut self, new:f32, viewbox:&Viewbox){ 
                    self.x = new;
                    self.compute_extremities(true, None, None);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //y
                fn get_y(&self) -> f32 { self.y }
                fn set_y(&mut self, new:f32, viewbox:&Viewbox){ 
                    self.y = new;
                    self.compute_extremities(true, None, None);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //angle
                fn get_angle(&self) -> f32 { self.angle }
                fn set_angle(&mut self, new:f32, viewbox:&Viewbox) { 
                    self.angle = new;
                    self.compute_extremities(true, None, None);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }
            //scale
                fn get_scale(&self) -> f32 { self.scale }
                fn set_scale(&mut self, new:f32, viewbox:&Viewbox) { 
                    self.scale = new;
                    self.compute_extremities(true, None, None);
                    self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                    self.request_render();
                }

        //other
            fn get_ignored(&self) -> bool { self.ignored }
            fn set_ignored(&mut self, new:bool) { self.ignored = new; }

        //offset
            fn get_cached_offset(&self) -> &Offset { &self.cached_offset }
            fn get_cached_offset_mut(&mut self) -> &mut Offset { &mut self.cached_offset }
            fn set_cached_offset(&mut self, new:Offset) { self.cached_offset = new; }

        //heed camera
            fn get_cached_heed_camera(&self) -> bool { self.cached_heed_camera }
            fn set_cached_heed_camera(&mut self, new:bool) { self.cached_heed_camera = new; }

        //extremities
            fn get_extremities(&self) -> &SimplePolygon { &self.extremities }
            fn __set_extremities(&mut self, new:SimplePolygon) { self.extremities = new; }

        //render
            //visibility
                fn is_visible(&self) -> bool { self.is_visible }
                fn set_is_visible(&mut self, new:bool) { 
                    self.previous_is_visible = self.is_visible;
                    self.is_visible = new;
                }
                fn previous_is_visible(&self) -> bool { self.previous_is_visible }
            //dot frame
                fn get_dot_frame(&self) -> bool { self.dot_frame }
                fn set_dot_frame(&mut self, new:bool) { self.dot_frame = new; }

    //element specific
        //casting
            fn as_character(&self) -> Option<&Character> { Some(self) }
            fn as_character_mut(&mut self) -> Option<&mut Character> { Some(self) }

        //universal attribute
            fn set_unified_attribute_from_js_sys_object(
                &mut self,
                unified_attribute: js_sys::Object, 
                font_requester: &RefCell<FontRequester>, 
                worker: &web_sys::Worker, 
                viewbox: &Viewbox
            ) {
                let printing_mode_object = match get_js_sys_object_from_js_sys_object("printingMode", &unified_attribute, true) {
                    Some(a) => a,
                    None => js_sys::Object::new(),
                };

                self.set_unified_attribute(
                    worker,
                    get_value_from_object__f32("x", &unified_attribute, true),
                    get_value_from_object__f32("y", &unified_attribute, true),
                    get_value_from_object__f32("angle", &unified_attribute, true),
                    get_value_from_object__f32("scale", &unified_attribute, true),
                    get_value_from_object__f32("width", &unified_attribute, true),
                    get_value_from_object__f32("height", &unified_attribute, true),
                    get_value_from_object__point("anchor", &unified_attribute, true),
                    get_value_from_object__string("font", &unified_attribute, true),
                    get_value_from_object__char("character", &unified_attribute, true),
                    match get_value_from_object__string("horizontal", &printing_mode_object, true){
                        None => None, 
                        Some(a) => PrintingModePositionHorizontal::from_str(&a),
                    },
                    match get_value_from_object__string("vertical", &printing_mode_object, true){
                        None => None, 
                        Some(a) => PrintingModePositionVertical::from_str(&a),
                    },
                    get_value_from_object__colour("colour", &unified_attribute, true),
                    font_requester,
                    viewbox,
                );
            }

        //point calculation
            fn calculate_points(&mut self, font_requester:&RefCell<FontRequester>, _worker:&web_sys::Worker, viewbox:&Viewbox) {
                self.vao_points = font_requester.borrow_mut().request_font(&self.font_name, self.id).get_vector(self.character).to_vec();

                //adjust for vertical printingMode
                    let mut horizontal_adjust = font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_right();
                    if self.printing_mode_horizontal == PrintingModePositionHorizontal::Middle { horizontal_adjust = horizontal_adjust/2.0; }
                    if self.printing_mode_horizontal != PrintingModePositionHorizontal::Left {
                        for index in 0..(self.vao_points.len()/2) {
                            self.vao_points[index*2] = self.vao_points[index*2] - horizontal_adjust;
                        }
                    }
                
                //adjust for vertical printingMode
                    let mut vertical_adjust = font_requester.borrow_mut().request_font(&self.font_name, self.id).get_misc_data(self.character).get_top();
                    if self.printing_mode_vertical == PrintingModePositionVertical::Middle { vertical_adjust = vertical_adjust/2.0; }
                    if self.printing_mode_vertical != PrintingModePositionVertical::Bottom {
                        for index in 0..(self.vao_points.len()/2) {
                            self.vao_points[index*2 + 1] = self.vao_points[index*2 + 1] - vertical_adjust;
                        }
                    }

                self.points_changed = true;
                self.compute_extremities(true, None, None);
                self.determine_if_visible(self.get_parent_clipping_polygon().as_ref(), viewbox, true);
                self.request_render();
            }

        //extremities
            fn get_length_of_points_for_extremity_calculation(&self) -> usize {
                self.vao_points.len()/2
            }
            fn get_point_for_extremity_calculation(&self, index:usize) -> Option<(f32,f32)> {
                if index >= self.vao_points.len()/2 {
                    return None;
                }

                Some( (
                    self.width * (self.vao_points[index*2 ] - self.anchor.get_x()),
                    self.height * (self.vao_points[index*2 +1] - self.anchor.get_y()),
                ) )
            }

        //render
            fn activate_webGL_render(
                &mut self, 
                offset: Option<&Offset>,
                context: &WebGl2RenderingContext, 
                web_gl2_program_conglomerate_manager: &mut WebGl2programConglomerateManager,
                _image_requester: &mut ImageRequester,
                resolution: &(u32, u32),
                stats: &mut Stats,
            ) -> bool {
                //vao
                    if self.vao_id.is_none() {
                        self.vao_id = Some( web_gl2_program_conglomerate_manager.generate_new_VAO_id(&self.element_type) );
                    }
        
                    if self.points_changed {
                        if web_gl2_program_conglomerate_manager.program_is_loaded(self.element_type) {
                            web_gl2_program_conglomerate_manager.update_VAO(
                                &context,
                                self.element_type,
                                self.vao_id.unwrap(),
                                &VAO_NAMES,
                                &[&self.vao_points],
                                &VAO_GROUPING_SIZE,
                            );
                        }
                    }
                    self.points_changed = false;
    
                //load program
                    web_gl2_program_conglomerate_manager.load_program(
                        &context,
                        Some(self.element_type),
                        &VERTEX_SHADER_SOURCE,
                        &FRAGMENT_SHADER_SOURCE,
                        self.vao_id.unwrap(),
                        &VAO_NAMES,
                        &[&self.vao_points],
                        &VAO_GROUPING_SIZE,
                        &UNIFORMS_TO_FIND,
                        None,
                    );

                //determine which offset to use
                    let working_offset = match offset {
                        None => self.get_cached_offset(),
                        Some(offset) => { offset },
                    };
                    
                //load uniforms
                    let uniforms = web_gl2_program_conglomerate_manager.get_uniform_locations(self.element_type).unwrap();
                    context.uniform2f(Some(&uniforms[0]), working_offset.get_x(), working_offset.get_y());
                    context.uniform1f(Some(&uniforms[1]), working_offset.get_scale());
                    context.uniform1f(Some(&uniforms[2]), working_offset.get_angle());
                    context.uniform2f(Some(&uniforms[3]), resolution.0 as f32, resolution.1 as f32);
                    context.uniform2f(Some(&uniforms[4]), self.width, self.height);
                    context.uniform2f(Some(&uniforms[5]), self.anchor.get_x(), self.anchor.get_y());
                    context.uniform4f(Some(&uniforms[6]), self.colour.premultiplied_r(), self.colour.premultiplied_g(), self.colour.premultiplied_b(), self.colour.a());
    
                //activate draw
                    context.draw_arrays(WebGl2RenderingContext::TRIANGLES, 0, (self.vao_points.len()/2) as i32);

                if stats.get_active() { stats.element_render_decision_register_info(self.get_id(), self.get_element_type(), RenderDecision::Rendered); }
                false
            }

        //info/dump
            fn _specific_info(&self) -> String {
                format!(
                    "width:{}, height:{}, anchor:{}, font_name:{}, character:{}, printing_mode_horizontal:{}, printing_mode_vertical:{}, colour:{}",
                    self.width,
                    self.height,
                    self.anchor,
                    self.font_name,
                    self.character,
                    self.printing_mode_horizontal,
                    self.printing_mode_vertical,
                    self.colour,
                )
            }
}